csrc/ir_utils.cpp

```diff
@@ -572,6 +574,45 @@ bool isReductionTvOp(const Expr* expr) {
   return ir_utils::isTvOp(expr) && isReductionOp(expr);
 }
 
+bool isReductionOverSizeZero(
+    const Expr* expr,
+    SchedulerRuntimeInfo& runtime_info) {
+  if (!isReductionOp(expr)) {
+    return false;
+  }
+  auto expr_eval = runtime_info.expressionEvaluator();
+  for (auto output : ir_utils::filterByType<TensorView>(expr->outputs())) {
+    for (auto id : output->getRootDomain()) {
+      if (id->isReduction() &&
+          expr_eval.evaluate(id->extent())->as<int64_t>() == 0) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void replaceReductionWithFull(Expr* expr) {
+  if (expr->isA<ReductionOp>()) {
+    // Fill value is reduction starting value
+    auto rop = expr->as<ReductionOp>();
+    auto out_root = rop->out()->as<TensorView>()->domain()->root();
+    auto out_shape = std::vector<Val*>(out_root.size());
+    for (auto i : c10::irange(out_root.size())) {
+      out_shape[i] = out_root[i]->extent();
+    }
+    auto new_out = full(out_shape, rop->init(), rop->out()->dtype());
+    ir_utils::replaceValInExpr(expr, rop->out(), new_out);
```

Comment:
We don't want to use `expr` anymore as it's a reduction op.

Instead, try this:

```
auto rop_out = rop->out();
auto out_uses = rop_out->uses();
for (auto out_use: out_uses) {
  ir_utils::replaceValInExpr(out_use, rop_out, new_out);
}
// At this point, rop_out is no longer used by any expr. 
rop->container()->removeVal(rop);
```

I believe this should remove the original reduction expr and its output from the fusion. It should also trigger updating the usage map, so the input should become unused.

Let me know if this works.

Reply:
Yeah you're right. I pushed a recursive version that removes the branch. After that the test passes. I am now extending this to Welford in which case instead of full we have for example inAvg / inN, which may be nan, matching pytorch.

Reply:
Just want to understand, is this recursive removal necessary?

Reply:
I think it's necessary, since there could be ops before the reduction. For example, the test currently reads:
```c++
  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);
  auto tv2 = sum(tv0, {1});
  fusion->addOutput(tv2);
```
Removing the sum makes it so that `tv0->uses()` is empty so that it doesn't complain here: https://github.com/NVIDIA/Fuser/blob/main/csrc/scheduler/pointwise_utils.cpp#L109

If we have another op before the reduction:
```c++
  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);
  auto tv1 = neg(tv0);
  auto tv2 = sum(tv1, {1});
  fusion->addOutput(tv2);
```
and we just remove the sum, `tv0` would still have a use, so it would block `tv2` being a valid reference tensor.

Reply:
It shouldn't be. If the sum is removed, the fusion would look like:

```
  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);
  auto tv1 = neg(tv0);
  auto tv3 = full(...);
  fusion->addOutput(tv3);
```

Then, the use map should be updated and the neg should no longer appear as a use of `tv0` since it's not connected with the output. I think its allocated objects are still kept in the fusion container, so removing them is helpful, but shouldn't be necessary.

Reply:
Your right that `resetTvUses()` updates uses for us, as you say. I removed the prune function. I am now trying to fix another issue that came up when I added more ops: finding reference tensor looks at all input IterDomains to see that they map to the reference, but since we accept reductions now those reduced IterDomains in the input are  not mapped to the reference. So I'm updating `DomainMap` to make exceptions for reduced domains in the input.